### 2 задание

In [1]:
import re
import numpy
import hashlib

In [2]:
p = 0.1 #P(false positive)
m = 0 #размер массива
n = 0 #кол-во элементов
k = 0 #optimal number of hash functions


In [3]:
def words(text):            #метод для обработки текста, очищаем от знаков препинания и разделяем по словам
    words = []
    with open(text, "r", encoding="utf-8") as file: 
        for line in file.readlines():
            string = re.sub("[^\w\s]", "", line)
            words_in_line = string.split()
            words = words + words_in_line
        return words


In [4]:
n = len(words('text.txt'))  #кол-во слов
n

94

In [5]:
m = round(-((n*numpy.log(p))/((numpy.log(2))**2)))   #формула для вычисления длины массива для фильтра
m

450

In [6]:
k = round(m/n*(numpy.log(2)))        #формула для оптимального кол-ва хэш-функций
k

3

In [7]:
array = [0] * m      #массив для фильтра

In [8]:
def hashing(element):                #метод для хэширования
    keys = []
    hash1 = hashlib.md5(element.encode()) 
    hash2 = hashlib.sha1(element.encode())
    hash3 = hashlib.sha224(element.encode())
    keys.append(int(hash1.hexdigest(), 16))  
    keys.append(int(hash2.hexdigest(), 16))
    keys.append(int(hash3.hexdigest(), 16))
    return keys

In [9]:
def add(element):                 #метод для добавления слов в фильтр
    key = 0 
    keys = hashing(element)
    for i in range(k): 
        key = int(keys[i] % m) 
        array[key] += 1  

In [10]:
for i in range(n):              #добавляем элементы
    add(words('text.txt')[i])

##### Использовав формулы, я выяснила, что оптимальный размер массива для фильтра = 450, оптимальное кол-во хэш-функций = 3
